File: modules/01b-ml-foundations/ts/04-roc-kmeans.ts
/**
 * Task 4 🟢 — Ranking metrics (ROC/AUC) and clustering (k-means) from scratch.
 *
 * What you'll learn:
 *   Part A — ROC curves and AUC:
 *     - A classifier outputs SCORES, not just labels; a threshold turns scores
 *       into decisions. Sweeping the threshold traces the ROC curve.
 *     - TPR (recall) vs FPR at every threshold; AUC summarises the whole curve.
 *     - AUC = P(a random positive scores higher than a random negative). It is
 *       threshold-free, which is why interviewers love it.
 *
 *   Part B — k-means (Lloyd's algorithm):
 *     - Unsupervised clustering: no labels, just group points by proximity.
 *     - Alternate: (1) assign each point to its nearest centroid, (2) move each
 *       centroid to the mean of its assigned points. Repeat until stable.
 *     - Inertia (within-cluster sum of squared distances) never increases.
 *
 * The math (README derives each step):
 *
 *   ROC:  at threshold t, predict positive iff score ≥ t.
 *         TPR(t) = TP / (TP + FN)     FPR(t) = FP / (FP + TN)
 *   AUC:  area under the (FPR, TPR) curve via the trapezoidal rule:
 *         AUC = Σ (fpr[i+1] - fpr[i]) · (tpr[i+1] + tpr[i]) / 2
 *
 *   k-means:
 *         assign:  cluster(x) = argmin_k ||x - c_k||²
 *         update:  c_k = mean of all x assigned to k
 *         inertia: Σ_i ||x_i - c_{assign(i)}||²
 *
 * You implement: rocCurve, auc, assignClusters, updateCentroids, inertia — plain
 * arrays only. Everything else is provided and runnable.
 *
 * How to run:
 *   pnpm tsx modules/01b-ml-foundations/ts/04-roc-kmeans.ts
 */

const SEED = 5;

// ---------------------------------------------------------------------------
// Seeded RNG (provided) — LCG + Box-Muller
// ---------------------------------------------------------------------------

function makeRng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296;
  };
}

function makeGaussian(seed: number): () => number {
  const u = makeRng(seed);
  return () => {
    const u1 = u() + 1e-12;
    const u2 = u();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  };
}

// ===========================================================================
// PART A — ROC / AUC  (YOU implement rocCurve and auc)
// ===========================================================================

/**
 * Compute the ROC curve by sweeping every threshold.
 *
 * Args:
 *   scores : the classifier's confidence that each item is positive
 *   labels : ground truth in {0, 1}
 *
 * Returns { fpr, tpr }: parallel arrays. The curve starts at (0, 0) and ends at
 * (1, 1).
 *
 * Method:
 *   1. Sort item INDICES by score DESCENDING (highest score first).
 *   2. P = count of positives, Nn = count of negatives.
 *   3. Start at (FPR=0, TPR=0). Walk the sorted order; on each positive admitted
 *      tp += 1, on each negative fp += 1; after each step push (fp/Nn, tp/P).
 *
 * TODO: implement.
 *   1. const order = labels.map((_, i) => i).sort((a, b) => scores[b] - scores[a]);
 *   2. const P = labels.filter(l => l === 1).length;
 *      const Nn = labels.filter(l => l === 0).length;
 *   3. let tp = 0, fp = 0;
 *      const fpr = [0], tpr = [0];
 *      for (const i of order) {
 *        if (labels[i] === 1) tp++; else fp++;
 *        tpr.push(tp / P);
 *        fpr.push(fp / Nn);
 *      }
 *   4. return { fpr, tpr };
 */
function rocCurve(
  scores: number[],
  labels: number[],
): { fpr: number[]; tpr: number[] } {
  // TODO: implement the ROC sweep
  throw new Error("TODO: implement rocCurve()");
}

/**
 * Area under the ROC curve via the trapezoidal rule.
 *
 * AUC = Σ_i (fpr[i+1] - fpr[i]) · (tpr[i+1] + tpr[i]) / 2
 *
 * TODO: implement.
 *   let area = 0;
 *   for (let i = 0; i < fpr.length - 1; i++) {
 *     area += (fpr[i + 1] - fpr[i]) * (tpr[i + 1] + tpr[i]) / 2;
 *   }
 *   return area;
 */
function auc(fpr: number[], tpr: number[]): number {
  // TODO: implement the trapezoidal AUC
  throw new Error("TODO: implement auc()");
}

// ===========================================================================
// PART B — k-means  (YOU implement assignClusters, updateCentroids, inertia)
// ===========================================================================

/** Squared Euclidean distance between two equal-length points. */
function sqDist(a: number[], b: number[]): number {
  let s = 0;
  for (let d = 0; d < a.length; d++) s += (a[d] - b[d]) ** 2;
  return s;
}

/**
 * Assign each point to its nearest centroid (by squared Euclidean distance).
 *
 * Returns an array of length X.length: the cluster index of each point.
 *
 * TODO: implement.
 *   return X.map((x) => {
 *     let best = 0, bestD = Infinity;
 *     for (let j = 0; j < centroids.length; j++) {
 *       const d = sqDist(x, centroids[j]);   // sqDist is provided above
 *       if (d < bestD) { bestD = d; best = j; }
 *     }
 *     return best;
 *   });
 */
function assignClusters(X: number[][], centroids: number[][]): number[] {
  // TODO: implement nearest-centroid assignment
  throw new Error("TODO: implement assignClusters()");
}

/**
 * Recompute each centroid as the mean of the points assigned to it.
 *
 * Returns k centroids. (For this seeded data no cluster goes empty; if one did,
 * leaving it as zeros / its old value is a reasonable fallback.)
 *
 * TODO: implement.
 *   const D = X[0].length;
 *   const sums = Array.from({ length: k }, () => new Array(D).fill(0));
 *   const counts = new Array(k).fill(0);
 *   for (let i = 0; i < X.length; i++) {
 *     const c = assignments[i];
 *     counts[c]++;
 *     for (let d = 0; d < D; d++) sums[c][d] += X[i][d];
 *   }
 *   return sums.map((s, c) => (counts[c] > 0 ? s.map((v) => v / counts[c]) : s));
 */
function updateCentroids(X: number[][], assignments: number[], k: number): number[][] {
  // TODO: implement the centroid update
  throw new Error("TODO: implement updateCentroids()");
}

/**
 * Within-cluster sum of squared distances (the k-means objective).
 *
 * inertia = Σ_i ||x_i - c_{assignments[i]}||²
 *
 * TODO: implement.
 *   let total = 0;
 *   for (let i = 0; i < X.length; i++) total += sqDist(X[i], centroids[assignments[i]]);
 *   return total;
 */
function inertia(X: number[][], centroids: number[][], assignments: number[]): number {
  // TODO: implement the inertia
  throw new Error("TODO: implement inertia()");
}

// ---------------------------------------------------------------------------
// Data + k-means driver (provided — do not edit)
// ---------------------------------------------------------------------------

/** Ground-truth labels + a realistic 'good but imperfect' score for each item. */
function makeScores(): { scores: number[]; labels: number[] } {
  const g = makeGaussian(SEED);
  const labels: number[] = [];
  const scores: number[] = [];
  for (let i = 0; i < 50; i++) {
    labels.push(1);
    scores.push(1.0 + g()); // positives score higher on average
  }
  for (let i = 0; i < 50; i++) {
    labels.push(0);
    scores.push(-1.0 + g());
  }
  return { scores, labels };
}

/** Three well-separated 2-D Gaussian blobs. Returns points + true labels. */
function makeBlobs(): { X: number[][]; trueLabel: number[] } {
  const g = makeGaussian(SEED + 7);
  const centers = [
    [0.0, 0.0],
    [6.0, 6.0],
    [0.0, 6.0],
  ];
  const X: number[][] = [];
  const trueLabel: number[] = [];
  centers.forEach((ctr, cls) => {
    for (let i = 0; i < 60; i++) {
      X.push([ctr[0] + 0.7 * g(), ctr[1] + 0.7 * g()]);
      trueLabel.push(cls);
    }
  });
  return { X, trueLabel };
}

/**
 * Deterministic k-means++-style init (farthest-first traversal): pick a seeded
 * first centre, then each next centre is the point farthest (by squared
 * distance) from the centres chosen so far. This spreads the initial centroids
 * across the blobs and avoids the "two centres land in one blob" local optimum
 * that plain random init can fall into. Returns k distinct point indices.
 */
function chooseInit(X: number[][], k: number, seed: number): number[] {
  const rng = makeRng(seed);
  const first = Math.floor(rng() * X.length);
  const chosen = [first];
  while (chosen.length < k) {
    let bestIdx = 0;
    let bestDist = -1;
    for (let i = 0; i < X.length; i++) {
      if (chosen.includes(i)) continue;
      // distance to the NEAREST already-chosen centre
      let nearest = Infinity;
      for (const c of chosen) nearest = Math.min(nearest, sqDist(X[i], X[c]));
      if (nearest > bestDist) {
        bestDist = nearest;
        bestIdx = i;
      }
    }
    chosen.push(bestIdx);
  }
  return chosen;
}

/** Full Lloyd's algorithm using your assign/update/inertia. Provided. */
function kmeans(
  X: number[][],
  k: number,
  maxIter = 50,
  seed = SEED,
): { centroids: number[][]; assignments: number[]; history: number[] } {
  const initIdx = chooseInit(X, k, seed);
  let centroids = initIdx.map((i) => [...X[i]]);

  let assignments = assignClusters(X, centroids);
  const history = [inertia(X, centroids, assignments)];

  for (let iter = 0; iter < maxIter; iter++) {
    centroids = updateCentroids(X, assignments, k);
    const newAssign = assignClusters(X, centroids);
    history.push(inertia(X, centroids, newAssign));
    const same = newAssign.every((v, i) => v === assignments[i]);
    assignments = newAssign;
    if (same) break;
  }

  return { centroids, assignments, history };
}

// ---------------------------------------------------------------------------
// Harness (provided)
// ---------------------------------------------------------------------------

function main(): void {
  console.log("Task 4 — ROC/AUC and k-means from scratch\n");

  // ── PART A: ROC / AUC ──────────────────────────────────────────────────────
  console.log("=".repeat(60));
  console.log("PART A — ROC / AUC");
  console.log("=".repeat(60));

  const { scores, labels } = makeScores();

  const perfect = labels.map((l) => l); // score == label → perfect ranking
  const reversedRanker = labels.map((l) => 1 - l); // exactly wrong
  const rg = makeGaussian(SEED + 3);
  const randomRanker = labels.map(() => rg()); // unrelated to labels

  const aucOf = (s: number[]): number => {
    const { fpr, tpr } = rocCurve(s, labels);
    return auc(fpr, tpr);
  };

  const aucPerfect = aucOf(perfect);
  const aucReversed = aucOf(reversedRanker);
  const aucRandom = aucOf(randomRanker);
  const aucReal = aucOf(scores);

  console.log(`\n  AUC (perfect ranker)  = ${aucPerfect.toFixed(3)}   (expect 1.0)`);
  console.log(`  AUC (reversed ranker) = ${aucReversed.toFixed(3)}   (expect 0.0)`);
  console.log(`  AUC (random ranker)   = ${aucRandom.toFixed(3)}   (expect ≈ 0.5)`);
  console.log(
    `  AUC (realistic model) = ${aucReal.toFixed(3)}   (good separation → high)`,
  );

  const { fpr, tpr } = rocCurve(scores, labels);
  console.log(
    `\n  ROC curve has ${fpr.length} points; starts at (${fpr[0].toFixed(1)}, ${tpr[0].toFixed(1)})` +
      ` ends at (${fpr[fpr.length - 1].toFixed(1)}, ${tpr[tpr.length - 1].toFixed(1)}).`,
  );

  // ── PART B: k-means ────────────────────────────────────────────────────────
  console.log("\n" + "=".repeat(60));
  console.log("PART B — k-means (Lloyd's algorithm)");
  console.log("=".repeat(60));

  const { X, trueLabel } = makeBlobs();
  const k = 3;
  const { assignments, history } = kmeans(X, k);

  console.log(`\n  Ran k-means (k=${k}) on ${X.length} points in 3 true blobs.`);
  console.log(`  Iterations recorded: ${history.length}`);
  console.log("  Inertia over iterations:");
  history.forEach((val, i) => {
    console.log(
      `    iter ${String(i).padStart(2)}: inertia = ${val.toFixed(2).padStart(10)}`,
    );
  });

  let nonIncreasing = true;
  for (let i = 0; i < history.length - 1; i++)
    if (history[i + 1] > history[i] + 1e-6) nonIncreasing = false;

  // Cluster recovery: each TRUE blob should be dominated by one cluster id.
  const dominantIds: number[] = [];
  let recovered = 0;
  for (let cls = 0; cls < k; cls++) {
    const counts = new Array(k).fill(0);
    let total = 0;
    for (let i = 0; i < trueLabel.length; i++) {
      if (trueLabel[i] === cls) {
        counts[assignments[i]]++;
        total++;
      }
    }
    let best = 0;
    for (let j = 1; j < k; j++) if (counts[j] > counts[best]) best = j;
    dominantIds.push(best);
    if (counts[best] / total >= 0.9) recovered++;
  }
  const distinct = new Set(dominantIds).size === k;

  console.log(`\n  Inertia non-increasing every iteration: ${nonIncreasing}`);
  console.log(`  True blobs cleanly recovered (≥90% one cluster): ${recovered}/${k}`);
  console.log(
    `  Dominant cluster ids per blob: [${dominantIds}]  (distinct: ${distinct})`,
  );

  // ── Acceptance checks ──────────────────────────────────────────────────────
  console.log("\nAcceptance:");
  const okPerfect = Math.abs(aucPerfect - 1.0) < 1e-9;
  const okReversed = Math.abs(aucReversed - 0.0) < 1e-9;
  const okRandom = Math.abs(aucRandom - 0.5) < 0.15;
  const okNonInc = nonIncreasing;
  const okRecover = recovered === k && distinct;

  console.log(
    `  [${okPerfect ? "x" : " "}] AUC of perfect ranker == 1.0  (${aucPerfect.toFixed(3)})`,
  );
  console.log(
    `  [${okReversed ? "x" : " "}] AUC of reversed ranker == 0.0  (${aucReversed.toFixed(3)})`,
  );
  console.log(
    `  [${okRandom ? "x" : " "}] AUC of random ranker ≈ 0.5 (±0.15)  (${aucRandom.toFixed(3)})`,
  );
  console.log(
    `  [${okNonInc ? "x" : " "}] k-means inertia non-increasing each iteration`,
  );
  console.log(
    `  [${okRecover ? "x" : " "}] k-means recovers all ${k} blobs (distinct clusters)`,
  );

  if (okPerfect && okReversed && okRandom && okNonInc && okRecover) {
    console.log("\n  All acceptance checks passed.");
  } else {
    console.log("\n  Some checks failed — revisit your implementation.");
  }
}

main();

// This file has no imports; the empty export marks it as an ES module so its
// top-level names stay file-scoped (keeps `pnpm typecheck` happy across modules).
export {};

File: modules/01b-ml-foundations/py/04_roc_kmeans.py
"""
Task 4 🟢 — Ranking metrics (ROC/AUC) and clustering (k-means) from scratch.

What you'll learn:
  Part A — ROC curves and AUC:
    - A classifier outputs SCORES, not just labels; a threshold turns scores into
      decisions. Sweeping the threshold traces the ROC curve.
    - TPR (recall) vs FPR at every threshold; AUC summarises the whole curve.
    - AUC = P(a random positive scores higher than a random negative). It is
      threshold-free, which is why interviewers love it.

  Part B — k-means (Lloyd's algorithm):
    - Unsupervised clustering: no labels, just group points by proximity.
    - Alternate two steps until convergence: (1) assign each point to its nearest
      centroid, (2) move each centroid to the mean of its assigned points.
    - Inertia (within-cluster sum of squared distances) never increases — that's
      why the algorithm is guaranteed to converge.

The math (README derives each step):

  ROC:  at threshold t, predict positive iff score ≥ t.
        TPR(t) = TP / (TP + FN)     (fraction of real positives caught)
        FPR(t) = FP / (FP + TN)     (fraction of real negatives wrongly flagged)
        Sweep t from high→low; each unique score is a candidate threshold.
  AUC:  area under the (FPR, TPR) curve, computed with the trapezoidal rule:
        AUC = Σ (fpr[i+1] - fpr[i]) · (tpr[i+1] + tpr[i]) / 2

  k-means:
        assign:  cluster(x) = argmin_k ||x - c_k||²
        update:  c_k = mean of all x assigned to k
        inertia: Σ_i ||x_i - c_{assign(i)}||²

You implement: roc_curve, auc, assign_clusters, update_centroids, and inertia.
Everything else — data, the three sanity rankers, the k-means loop, and the
cluster-recovery check — is provided and runnable.

How to run:
  uv run python modules/01b-ml-foundations/py/04_roc_kmeans.py
"""

from __future__ import annotations

import numpy as np

SEED = 5


# ===========================================================================
# PART A — ROC / AUC
# ===========================================================================


def roc_curve(scores: np.ndarray, labels: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """
    Compute the ROC curve by sweeping every threshold.

    Args:
      scores : (N,) float — the classifier's confidence that each item is positive
      labels : (N,) int in {0, 1} — the ground truth

    Returns:
      fpr : (M,) false-positive rates, ascending, starting at 0 and ending at 1
      tpr : (M,) true-positive  rates, matched to fpr

    Method (a clean, standard way):
      1. P = number of positives (labels == 1); Nn = number of negatives.
      2. Sort the items by score DESCENDING. As you lower the threshold you admit
         items one at a time in that order.
      3. Start at the point (FPR=0, TPR=0) — threshold above every score.
      4. Walk down the sorted list; each time you admit a positive, TP += 1; each
         time you admit a negative, FP += 1. After each step record
         (FPR = FP / Nn, TPR = TP / P).
      5. The curve therefore starts at (0, 0) and ends at (1, 1).

    TODO: implement.
      1. order = np.argsort(-scores)               # indices, highest score first
      2. y = labels[order]
      3. P = (labels == 1).sum(); Nn = (labels == 0).sum()
      4. tp = fp = 0
         tpr_list = [0.0]; fpr_list = [0.0]
         for lbl in y:
             if lbl == 1: tp += 1
             else:        fp += 1
             tpr_list.append(tp / P)
             fpr_list.append(fp / Nn)
      5. return np.array(fpr_list), np.array(tpr_list)
    """
    # TODO: implement the ROC sweep
    raise NotImplementedError("TODO: implement roc_curve()")


def auc(fpr: np.ndarray, tpr: np.ndarray) -> float:
    """
    Area under the ROC curve via the trapezoidal rule.

    AUC = Σ_i (fpr[i+1] - fpr[i]) · (tpr[i+1] + tpr[i]) / 2

    (fpr must be non-decreasing — roc_curve returns it that way.)

    TODO: implement.
      area = 0.0
      for i in range(len(fpr) - 1):
          area += (fpr[i+1] - fpr[i]) * (tpr[i+1] + tpr[i]) / 2
      return area
    (or, equivalently, return float(np.trapezoid(tpr, fpr)))
    """
    # TODO: implement the trapezoidal AUC
    raise NotImplementedError("TODO: implement auc()")


# ===========================================================================
# PART B — k-means (Lloyd's algorithm)
# ===========================================================================


def assign_clusters(X: np.ndarray, centroids: np.ndarray) -> np.ndarray:
    """
    Assign each point to its nearest centroid (by squared Euclidean distance).

    Args:
      X         : (N, D) points
      centroids : (k, D) current centroids

    Returns:
      (N,) int array — the cluster index of each point.

    TODO: implement.
      For each point x_i, compute ||x_i - c_j||² for every centroid j and take the
      argmin. A vectorised way:
        # dists[i, j] = squared distance from point i to centroid j
        diff = X[:, None, :] - centroids[None, :, :]   # (N, k, D)
        dists = (diff ** 2).sum(axis=2)                # (N, k)
        return np.argmin(dists, axis=1)                # (N,)
      (A plain double loop is fine too — clarity over cleverness.)
    """
    # TODO: implement nearest-centroid assignment
    raise NotImplementedError("TODO: implement assign_clusters()")


def update_centroids(X: np.ndarray, assignments: np.ndarray, k: int) -> np.ndarray:
    """
    Recompute each centroid as the mean of the points assigned to it.

    Args:
      X           : (N, D) points
      assignments : (N,) cluster index of each point
      k           : number of clusters

    Returns:
      (k, D) new centroids.

    Handle empty clusters: if no point is assigned to cluster j, keep a sensible
    fallback (e.g. a random point, or leave it as zeros). For this seeded data no
    cluster goes empty, so the simple mean is enough.

    TODO: implement.
      D = X.shape[1]
      new_centroids = np.zeros((k, D))
      for j in range(k):
          members = X[assignments == j]
          if len(members) > 0:
              new_centroids[j] = members.mean(axis=0)
      return new_centroids
    """
    # TODO: implement the centroid update
    raise NotImplementedError("TODO: implement update_centroids()")


def inertia(X: np.ndarray, centroids: np.ndarray, assignments: np.ndarray) -> float:
    """
    Within-cluster sum of squared distances (the k-means objective).

    inertia = Σ_i ||x_i - c_{assignments[i]}||²

    TODO: implement.
      total = 0.0
      for i in range(len(X)):
          diff = X[i] - centroids[assignments[i]]
          total += float(np.dot(diff, diff))
      return total
    (or vectorised: ((X - centroids[assignments]) ** 2).sum())
    """
    # TODO: implement the inertia
    raise NotImplementedError("TODO: implement inertia()")


# ---------------------------------------------------------------------------
# Data + harness  (provided — do not edit)
# ---------------------------------------------------------------------------


def make_scores() -> tuple[np.ndarray, np.ndarray]:
    """Ground-truth labels + a realistic 'good but imperfect' score for each item."""
    rng = np.random.default_rng(SEED)
    labels = np.concatenate([np.ones(50), np.zeros(50)]).astype(np.int64)
    # positives score higher on average, but the distributions overlap
    pos_scores = rng.normal(1.0, 1.0, 50)
    neg_scores = rng.normal(-1.0, 1.0, 50)
    scores = np.concatenate([pos_scores, neg_scores])
    return scores, labels


def make_blobs() -> tuple[np.ndarray, np.ndarray]:
    """Three well-separated 2-D Gaussian blobs. Returns (X, true_label)."""
    rng = np.random.default_rng(SEED)
    centers = np.array([[0.0, 0.0], [6.0, 6.0], [0.0, 6.0]])
    pts, true = [], []
    for cls, ctr in enumerate(centers):
        pts.append(rng.normal(ctr, 0.7, size=(60, 2)))
        true.append(np.full(60, cls))
    X = np.vstack(pts)
    true_label = np.concatenate(true)
    return X, true_label


def _init_centroids(X: np.ndarray, k: int, seed: int) -> np.ndarray:
    """
    Deterministic k-means++-style init (farthest-first traversal): a seeded first
    centre, then each next centre is the point farthest (by squared distance)
    from the centres chosen so far. This spreads the initial centroids across the
    blobs and avoids the "two centres land in one blob" local optimum that plain
    random init can fall into. Returns k point indices.
    """
    rng = np.random.default_rng(seed)
    chosen = [int(rng.integers(len(X)))]
    while len(chosen) < k:
        # distance of each point to its NEAREST already-chosen centre
        d = np.min(np.stack([((X - X[c]) ** 2).sum(axis=1) for c in chosen], axis=1), axis=1)
        d[chosen] = -1.0  # never re-pick an existing centre
        chosen.append(int(np.argmax(d)))
    return X[chosen].copy()


def kmeans(
    X: np.ndarray, k: int, max_iter: int = 50, seed: int = SEED
) -> tuple[np.ndarray, np.ndarray, list[float]]:
    """
    Full Lloyd's algorithm using your assign/update/inertia functions.

    Initialise centroids with a farthest-first traversal (see _init_centroids),
    then alternate assign/update until assignments stop changing or max_iter.
    Returns (centroids, assignments, inertia_history). Provided — don't edit.
    """
    centroids = _init_centroids(X, k, seed)

    assignments = assign_clusters(X, centroids)
    history = [inertia(X, centroids, assignments)]

    for _ in range(max_iter):
        centroids = update_centroids(X, assignments, k)
        new_assign = assign_clusters(X, centroids)
        history.append(inertia(X, centroids, new_assign))
        if np.array_equal(new_assign, assignments):
            assignments = new_assign
            break
        assignments = new_assign

    return centroids, assignments, history


def main() -> None:
    print("Task 4 — ROC/AUC and k-means from scratch\n")

    # ── PART A: ROC / AUC ────────────────────────────────────────────────────
    print("=" * 60)
    print("PART A — ROC / AUC")
    print("=" * 60)

    scores, labels = make_scores()

    # A perfect ranker: use the labels themselves as the score.
    perfect = labels.astype(float)
    # A reversed ranker: exactly wrong.
    reversed_ranker = 1.0 - labels.astype(float)
    # A random ranker: scores unrelated to labels.
    rng = np.random.default_rng(SEED + 5)
    random_ranker = rng.normal(0, 1, len(labels))

    def auc_of(s: np.ndarray) -> float:
        f, t = roc_curve(s, labels)
        return auc(f, t)

    auc_perfect = auc_of(perfect)
    auc_reversed = auc_of(reversed_ranker)
    auc_random = auc_of(random_ranker)
    auc_real = auc_of(scores)

    print(f"\n  AUC (perfect ranker)  = {auc_perfect:.3f}   (expect 1.0)")
    print(f"  AUC (reversed ranker) = {auc_reversed:.3f}   (expect 0.0)")
    print(f"  AUC (random ranker)   = {auc_random:.3f}   (expect ≈ 0.5)")
    print(f"  AUC (realistic model) = {auc_real:.3f}   (good separation → high)")

    fpr, tpr = roc_curve(scores, labels)
    print(
        f"\n  ROC curve has {len(fpr)} points; starts at "
        f"({fpr[0]:.1f}, {tpr[0]:.1f}) ends at ({fpr[-1]:.1f}, {tpr[-1]:.1f})."
    )

    # ── PART B: k-means ──────────────────────────────────────────────────────
    print("\n" + "=" * 60)
    print("PART B — k-means (Lloyd's algorithm)")
    print("=" * 60)

    X, true_label = make_blobs()
    k = 3
    centroids, assignments, history = kmeans(X, k)

    print(f"\n  Ran k-means (k={k}) on {len(X)} points in 3 true blobs.")
    print(f"  Iterations recorded: {len(history)}")
    print("  Inertia over iterations:")
    for i, val in enumerate(history):
        print(f"    iter {i:>2}: inertia = {val:>10.2f}")

    # Non-increasing inertia?
    non_increasing = all(history[i + 1] <= history[i] + 1e-6 for i in range(len(history) - 1))

    # Cluster recovery: each TRUE blob should be dominated by a single cluster id.
    recovered = 0
    for cls in range(k):
        mask = true_label == cls
        cluster_ids = assignments[mask]
        # most common cluster id among this blob's points
        counts = np.bincount(cluster_ids, minlength=k)
        dominant_frac = counts.max() / counts.sum()
        if dominant_frac >= 0.9:
            recovered += 1

    # Also confirm the 3 dominant cluster ids are distinct (no two blobs collapse).
    dominant_ids = []
    for cls in range(k):
        mask = true_label == cls
        dominant_ids.append(int(np.bincount(assignments[mask], minlength=k).argmax()))
    distinct = len(set(dominant_ids)) == k

    print(f"\n  Inertia non-increasing every iteration: {non_increasing}")
    print(f"  True blobs cleanly recovered (≥90% one cluster): {recovered}/{k}")
    print(f"  Dominant cluster ids per blob: {dominant_ids}  (distinct: {distinct})")

    # ── Acceptance checks ────────────────────────────────────────────────────
    print("\nAcceptance:")
    ok_perfect = abs(auc_perfect - 1.0) < 1e-9
    ok_reversed = abs(auc_reversed - 0.0) < 1e-9
    ok_random = abs(auc_random - 0.5) < 0.15
    ok_noninc = non_increasing
    ok_recover = recovered == k and distinct

    print(f"  [{'x' if ok_perfect else ' '}] AUC of perfect ranker == 1.0  ({auc_perfect:.3f})")
    print(f"  [{'x' if ok_reversed else ' '}] AUC of reversed ranker == 0.0  ({auc_reversed:.3f})")
    print(f"  [{'x' if ok_random else ' '}] AUC of random ranker ≈ 0.5 (±0.15)  ({auc_random:.3f})")
    print(f"  [{'x' if ok_noninc else ' '}] k-means inertia non-increasing each iteration")
    print(f"  [{'x' if ok_recover else ' '}] k-means recovers all {k} blobs (distinct clusters)")

    if ok_perfect and ok_reversed and ok_random and ok_noninc and ok_recover:
        print("\n  All acceptance checks passed.")
    else:
        print("\n  Some checks failed — revisit your implementation.")


if __name__ == "__main__":
    main()
